### PR TITLE
Laboratory work #2, Pletneva Polina - 19FPL1

### PR DESCRIPTION
**Pull request by standfromunder**
Кащихин Андрей

### CODE REVIEW

File: lab_1/sort_concordance_test.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from main import sort_concordance
 from main import tokenize
 from main import read_from_file
+from main import sort_concordance
```
Comment (WhiteJaeger) [inline lab_1/sort_concordance_test.py]: Remove.

File: lab_2/main.py [hunk 2]
```diff
@@ -34,7 +46,26 @@ def fill_lcs_matrix(first_sentence_tokens: tuple, second_sentence_tokens: tuple)
     :param second_sentence_tokens: a tuple of tokens
     :return: a lcs matrix
     """
-    pass
+    if not isinstance(first_sentence_tokens,tuple) or not isinstance(second_sentence_tokens, tuple)\
+        or not all(isinstance(word,str) for word in first_sentence_tokens)\
+        or not all(isinstance(word,str) for word in second_sentence_tokens):
+        return []
+    lcs_matrix = create_zero_matrix(len(first_sentence_tokens),len(second_sentence_tokens))
+    for i in range(len(first_sentence_tokens)):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `i` is not very descriptive name, right?

File: lab_2/main.py [hunk 3]
```diff
@@ -46,7 +77,13 @@ def find_lcs_length(first_sentence_tokens: tuple, second_sentence_tokens: tuple,
     :param plagiarism_threshold: a threshold
     :return: a length of the longest common subsequence
     """
-    pass
+    lcs_matrix = fill_lcs_matrix(first_sentence_tokens,second_sentence_tokens)
+    if lcs_matrix == [] or not isinstance(plagiarism_threshold,float) or plagiarism_threshold < 0 \
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `lcs_matrix == []` is it really needed? That explicit?

File: lab_2/main.py [hunk 4]
```diff
@@ -46,7 +77,13 @@ def find_lcs_length(first_sentence_tokens: tuple, second_sentence_tokens: tuple,
...
+            or plagiarism_threshold > 1:
+        return -1
+    if len(second_sentence_tokens) == 0 or len(lcs_matrix) / len(second_sentence_tokens) < plagiarism_threshold:
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `len(second_sentence_tokens) == 0` is it really needed to be that explicit?

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_2/main.py (hunk 3 above)
Comment (standfromunder) [reply lab_2/main.py]: why not? in function "fill_lcs_matrix" the same explicits for type of tokens

File: lab_2/main.py [hunk 5]
```diff
@@ -12,7 +12,12 @@ def tokenize_by_lines(text: str) -> tuple:
     e.g. text = 'I have a cat.\nHis name is Bruno'
     --> (('i', 'have', 'a', 'cat'), ('his', 'name', 'is', 'bruno'))
     """
-    pass
+    if not isinstance(text,str):
+        return ()
+    text = text.split('\n')
+    result = tuple(tuple(tokenize(x)) for x in text if tokenize(x))
```
Comment (WhiteJaeger) [inline lab_2/main.py]: You do `tokenize` call twice which is quite memory consuming.

File: lab_2/main.py [hunk 6]
```diff
@@ -34,7 +46,26 @@ def fill_lcs_matrix(first_sentence_tokens: tuple, second_sentence_tokens: tuple)
     :param second_sentence_tokens: a tuple of tokens
     :return: a lcs matrix
     """
-    pass
+    if not isinstance(first_sentence_tokens,tuple) or not isinstance(second_sentence_tokens, tuple)\
+        or not all(isinstance(word,str) for word in first_sentence_tokens)\
+        or not all(isinstance(word,str) for word in second_sentence_tokens):
+        return []
+    lcs_matrix = create_zero_matrix(len(first_sentence_tokens),len(second_sentence_tokens))
+    for index_1 in range(len(first_sentence_tokens)):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: Take a look at `enumerate`.

File: lab_2/main.py [hunk 7]
```diff
@@ -46,7 +77,20 @@ def find_lcs_length(first_sentence_tokens: tuple, second_sentence_tokens: tuple,
...
+    lcs_matrix = fill_lcs_matrix(first_sentence_tokens,second_sentence_tokens)
+    check_type = [isinstance(plagiarism_threshold,float), isinstance(first_sentence_tokens,tuple),
+                  isinstance(second_sentence_tokens, tuple)]
+    check_type_2 = False
+    if all(check_type):
+        check_type_2 = [all(isinstance(word, str) for word in second_sentence_tokens),
+                        all(isinstance(word, str) for word in first_sentence_tokens),
+                        plagiarism_threshold >= 0, plagiarism_threshold <= 1]
+    if all(check_type) and all(check_type_2) and\
+            (len(second_sentence_tokens) == 0 or len(lcs_matrix) / len(second_sentence_tokens) < plagiarism_threshold):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `len(second_sentence_tokens) == 0` can be shortened.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]

File: lab_2/main.py [hunk 8]
```diff
@@ -92,8 +189,31 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
+    if all(check_type):
+        check_type_2 = [all(isinstance(word, str) for word in original_sentence_tokens),
+                        all(isinstance(word, str) for word in suspicious_sentence_tokens),
+                        all(isinstance(word, str) for word in lcs)]
+    if not all(check_type) or not all(check_type_2):
+        return ()
+    diff_index_original = []
+    diff_index_suspicious = []
+    for index, word in enumerate(original_sentence_tokens):
+        if word not in lcs and (index == 0 or original_sentence_tokens[index - 1] in lcs):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `index == 0` con be shortened.

File: lab_2/main.py [hunk 9]
```diff
@@ -92,8 +189,31 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
+            diff_index_original.append(index)
+        if word not in lcs \
+                and (index == len(original_sentence_tokens) - 1 or original_sentence_tokens[index + 1] in lcs):
+            diff_index_original.append(index + 1)
+    for index, word in enumerate(suspicious_sentence_tokens):
+        if word not in lcs and (index == 0 or suspicious_sentence_tokens[index - 1] in lcs):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `index == 0` can be shortened.

**Review by WhiteJaeger — APPROVED** [review]
Good!
You are done here. Just fix little issues later.

**Pull request closed without merging**